src/deploy/functions/params.ts

```diff
@@ -0,0 +1,332 @@
+ */
+export function resolveBoolean(
+  from: boolean | build.Expression<boolean>,
+  paramValues: Record<string, build.Field<string | number | boolean>>
+): boolean {
+  if (typeof from === "string" && /{{ params\.(\w+) }}/.test(from)) {
+    const match = /{{ params\.(\w+) }}/.exec(from);
+    const referencedParamValue = paramValues[match![1]];
+    if (typeof referencedParamValue !== "boolean") {
+      throw new FirebaseError(
+        "Referenced boolean parameter '" +
+          match +
+          "' resolved to non-boolean value " +
+          referencedParamValue
+      );
+    }
+    return referencedParamValue;
+  } else if (typeof from === "string") {
+    throw new FirebaseError("CEL evaluation of expression '" + from + "' not yet supported");
+  }
+  return from;
+}
+
+interface ParamBase<T extends string | number | boolean> {
+  // name of the param. Will be exposed as an environment variable with this name
+  param: string;
+
+  // A human friendly name for the param. Will be used in install/configure flows to describe
+  // what param is being updated. If omitted, UX will use the value of "param" instead.
+  label?: string;
+
+  // A long description of the parameter's purpose and allowed values. If omitted, UX will not
+  // provide a description of the parameter
+  description?: string;
+
+  // Default value. If not provided, a param must be supplied.
+  default?: T | build.Expression<T>;
+
+  // default: false
+  immutable?: boolean;
+}
+
+export interface StringParam extends ParamBase<string> {
+  type: "string";
+
+  // If omitted, defaults to TextInput<string>
+  input?: TextInput<string> | SelectInput<string>;
+}
+
+export interface IntParam extends ParamBase<number> {
+  type: "int";
+
+  // If omitted, defaults to TextInput<number>
+  input?: TextInput<number> | SelectInput<number>;
+}
+
+export interface TextInput<T, Extensions = {}> { // eslint-disable-line
+  type: "text";
+
+  text:
+    | Extensions
+    | {
+        example?: string;
+      };
+}
+
+interface SelectOptions<T> {
+  // Optional human-facing value for this option (e.g. "US Central (Iowa)" instead of value
+  // "us-central1")
+  label?: string;
+
+  // Actual value of the parameter if this option is selected
+  value: T;
+}
+
+export interface SelectInput<T> {
+  type: "select";
+
+  select: Array<SelectOptions<T>>;
+}
+
+export type Param = StringParam | IntParam;
+type ParamValue = string | number | boolean;
+
+/**
+ * Calls the corresponding resolveX function for the type of a param.
+ * To be used when resolving the default value of a param, if CEL.
+ * It's an error to call this on a CEL expression that depends on params not already known in the currentEnv.
+ */
+function resolveDefaultCEL(
+  type: string,
+  expr: CEL,
+  currentEnv: Record<string, ParamValue>
+): ParamValue {
+  const deps = dependenciesCEL(expr);
+  const allDepsFound = deps.every((dep) => {
+    return Object.keys(currentEnv).includes(dep);
```

Comment:
nit: `hasOwnProperty` is oblique when called "properly" but is much faster here.

```
const allDepsFound = deps.every(dep => {}.hasOwnProperty.call(currentEnv, dep))
```

Also, I don't really care if you want to just test for truthiness since you know that currentEnv isn't going to have null values:

```
const allDepsFound = deps.every((dep) => !!currentEnv[dep])
```

(`!!` is a way to cast into a boolean to silence type concerns)

Reply:
Done.